projection/core/src/main/scala/com/lightbend/lagom/internal/projection/ProjectionRegistryActor.scala

```diff
@@ -4,128 +4,186 @@
+
+      // watch
       context.watch(sender)
 
-    case GetDesiredState =>
-      replicator ! Get(DataKey, ReadLocal, Some(sender()))
+    case GetState =>
+      sender ! State.fromReplicatedData(nameIndexLocalCopy, requestedStatusLocalCopy, observedStatusLocalCopy)
+
+    // StateRequestCommand come from `ProjectionRegistry` and contain a requested Status
+    case command: StateRequestCommand =>
+      // locate the target actor and send the request
+      log.debug(s"Propagating request $command.")
+      updateLWWMapForRequests(command.coordinates.asKey, command.requestedStatus)
+
+    // Bare Status come from worker and contain an observed Status
+    case observedStatus: Status =>
+      log.debug(s"Observed [${sender().path.toString}] as $observedStatus.")
+      reversedActorIndex.get(sender()) match {
+        case Some(workerName) => updateLWWMapForObserved(workerName, observedStatus)
+        case None             => log.error(s"Unknown actor [${sender().path.toString}] reports status $observedStatus.")
+      }
+
+    // TODO: handle UpdateSuccess/UpdateFailure !!
+    case UpdateSuccess(_, _) => //TODO
+    case _: UpdateFailure[_] => //TODO
+    case changed @ Changed(RequestedStatusDataKey) => {
+      val remotelyChanged                 = changed.get(RequestedStatusDataKey).entries
+      val diffs: Set[(WorkerKey, Status)] = remotelyChanged.toSet.diff(requestedStatusLocalCopy.toSet)
+
+      // when the requested status changes, we must forward the new value to the appropriate actor
+      // if it's a one of the workers in the local actorIndex
+      diffs
+        .foreach {
+          case (workerName, requestedStatus) =>
+            log.debug(s"Remotely requested worker [$workerName] as [$requestedStatus].")
+            actorIndex.get(workerName).foreach { workerRef =>
+              log.debug(
+                s"Setting requested status [$requestedStatus] on worker $workerName [${workerRef.path.toString}]"
+              )
+              workerRef ! requestedStatus
+            }
+        }
+
+      requestedStatusLocalCopy = remotelyChanged
+    }
+
+    // TODO: handle UpdateSuccess/UpdateFailure !!
+    case changed @ Changed(ObservedStatusDataKey) =>
+      observedStatusLocalCopy = changed.get(ObservedStatusDataKey).entries
 
-    case g @ GetSuccess(DataKey, req) =>
-      val registry: PNCounterMap[WorkerMetadata] = g.get(DataKey)
-      val desiredStatus: DesiredState            = mapStatus(registry.entries)
-      req.get.asInstanceOf[ActorRef] ! desiredStatus
+    // TODO: handle UpdateSuccess/UpdateFailure !!
+    case changed @ Changed(NameIndexDataKey) =>
+      nameIndexLocalCopy = changed.get(NameIndexDataKey).entries
 
     case Terminated(deadActor) =>
-      // update indices and stop watching
-      actorIndex = actorIndex - actorReverseIndex(deadActor)
-      actorReverseIndex = actorReverseIndex - deadActor
+      log.debug(s"Worker ${deadActor.path.name} died. Marking it as Stopped.")
```

Comment:
In which case, a WorkerHolder would stop?

Reply:
See above case of "down scale".

I don't think we should unregister and change the state.

When the WorkerHolder dies, it will restart anyway and it should retrieve the state it had before it stopped.

Reply:
See my comments above. Downscaling, doing a rolling upgrade or any op causing Workers to relocate causes a Worker to stop (albeit temporarily). When that happens, the _observed_ state must reflect the fact that the worker is indeed `Stopped`. It'll eventually come back to live somewhere else, correct, and only then it'll be observed as `Started` once again.

Reply:
got it, we are updating the observed state